keyboards/synthlabs/solo/keymaps/default/keymap.c

```diff
@@ -0,0 +1,36 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_all(
+                                       KC_J,    KC_U,    KC_K,    KC_I,    KC_L,    KC_O, KC_SCLN,
+        KC_VOLD, KC_MUTE, KC_VOLU,
+                                       KC_J,    KC_U,    KC_K,    KC_I,    KC_L,    KC_O, KC_SCLN
+    ),
+    [1] = LAYOUT_all(
+                                    _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______,
+                                    _______, _______, _______, _______, _______, _______, _______
+    ),
+    [2] = LAYOUT_all(
+                                    _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______,
+                                    _______, _______, _______, _______, _______, _______, _______
+    ),
+    [3] = LAYOUT_all(
+                                    _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______,
+                                    _______, _______, _______, _______, _______, _______, _______
+    )
+};
+
+const keypos_t keypos_cw  = { .row=1, .col=2 };
+const keypos_t keypos_ccw = { .row=1, .col=0 };
+
+uint16_t get_keycode_at_pos(keypos_t keypos) {
+    return keymap_key_to_keycode(layer_switch_get_layer(keypos), keypos);
+}
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    tap_code(get_keycode_at_pos(clockwise ? keypos_cw : keypos_ccw));
+    return true;
+}
```

Comment:
This isn't the right away to do this.

```suggestion
```

Reply:
I elected to give the rotary encoder keycodes a virtual position in the key matrix so that they can be remapped in VIA, which has been successful from my testing.

I haven't really seen this functionality elsewhere in the codebase - but maybe there's a better way to do it that I just missed. Any suggestions?

Reply:
The issue with this method is that it only supports basic keycodes.  Not even modded keycodes.  And it doesn't support anything that's not basic. 

If you want to do something like this, then this will work much better: 
```c
static uint8_t encoder_state[] = {0};
static keypos_t encoder_cw[] = ENCODERS_CW_KEY;
static keypos_t encoder_ccw[] = ENCODERS_CCW_KEY;
#endif

void encoder_action_unregister(void) {
#ifdef ENCODERS
    for (int index = 0; index < NUM_OF_ENCODERS; ++index) {
        if (encoder_state[index]) {
            keyevent_t encoder_event = (keyevent_t) {
                .key = encoder_state[index] >> 1 ? encoder_cw[index] : encoder_ccw[index],
                .pressed = false,
                .time = (timer_read() | 1)
            };
            encoder_state[index] = 0;
            action_exec(encoder_event);
        }
    }
}

void encoder_action_register(uint8_t index, bool clockwise) {
    keyevent_t encoder_event = (keyevent_t) {
        .key = clockwise ? encoder_cw[index] : encoder_ccw[index],
        .pressed = true,
        .time = (timer_read() | 1)
    };
    encoder_state[index] = (clockwise ^ 1) | (clockwise << 1);
    action_exec(encoder_event);
}

void matrix_scan_user(void) {
    encoder_action_unregister();
}

bool encoder_update_user(uint8_t index, bool clockwise) {
    encoder_action_register(index, clockwise);
    return false;
}
```

However, I would recommend this only for the via keymap, and add something like the following to the keyboard's c file: 

```c
#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

This way, a default behavior is enabled, but via has some custom handling.

Reply:
Ah I see, thanks for the suggestion! I'll look through this and test it out. Will update soon.

And for having this only in the via keycap, that makes perfect sense to me. Will do.

Reply:
Updated per @drashna's suggestions in 06e02994f184ebe537df9d5ab0e6eb34fe86bccd and cdad9cdf23e28190886bfd3d625e3776603eff10.